edge/poll.go

```diff
@@ -231,6 +250,53 @@ func (service *PollService) poll() error {
 	return service.processStacks(environmentStatus.Stacks)
 }
 
+func (service *PollService) processUpdate(versionUpdate edgetypes.VersionUpdateRequest) error {
+	// update is disabled or no new update
+	if !versionUpdate.Active || (versionUpdate.ScheduleID == service.versionUpdateStatus.ScheduleID && service.versionUpdateStatus.Status == edgetypes.UpdateScheduleStatusSuccess) {
+		log.Printf("[DEBUG] [edge] [message: no update available] [active: %t] [schedule_id: %d] [current_schedule_id: %d]", versionUpdate.Active, versionUpdate.ScheduleID, service.versionUpdateStatus.ScheduleID)
+		return nil
+	}
+
+	if versionUpdate.Active && service.versionUpdateStatus.Status == edgetypes.UpdateScheduleStatusSkip {
+		log.Printf("[DEBUG] [edge] [message: skipping update] [schedule_id: %d] [error: %s]", versionUpdate.ScheduleID, service.versionUpdateStatus.Error)
+		service.versionUpdateStatus.ScheduleID = versionUpdate.ScheduleID
+		return nil
+	}
+
+	// update is scheduled for the future
+	scheduledTime, err := time.Parse(edgetypes.UpdateScheduleTimeFormat, string(versionUpdate.ScheduledTime))
+	if err != nil {
+		return errors.WithMessage(err, "Unable to parse scheduled time")
+	}
+
+	if scheduledTime.After(time.Now()) {
+		log.Printf("[DEBUG] [edge] [message: update scheduled for the future] [scheduled_time: %v]", versionUpdate.ScheduledTime)
+		return nil
+	}
+
+	log.Printf("[DEBUG] [edge] [message: update available] [schedule_id: %d] [current_schedule_id: %d] [status: %d]", versionUpdate.ScheduleID, service.versionUpdateStatus.ScheduleID, service.versionUpdateStatus.Status)
+
+	service.versionUpdateStatus.ScheduleID = versionUpdate.ScheduleID
+	service.versionUpdateStatus.Error = ""
+	service.versionUpdateStatus.Status = 0
+
+	go func() {
+		service.Stop()
+	}()
+
+	defer service.Start()
```

Comment:
This statement races against `go service.Stop()`, why is it needed to stop it and start it again? The poll won't iterate again until this function returns anyway so I believe the `Start()` and `Stop()` calls can be removed.